powerauth-restful-security-javaee/src/main/java/io/getlime/security/powerauth/rest/api/jaxrs/controller/v3/TokenController.java

```diff
@@ -0,0 +1,182 @@
+
+                if (!"3.0".equals(authentication.getVersion())) {
+                    Logger.getLogger(this.getClass().getName()).log(Level.SEVERE, "Endpoint does not support PowerAuth protocol version {}", authentication.getVersion());
+                    throw new PowerAuthAuthenticationException();
+                }
+
+                if (!"3.0".equals(encryption.getVersion())) {
+                    Logger.getLogger(this.getClass().getName()).log(Level.SEVERE, "Endpoint does not support PowerAuth protocol version {}", encryption.getVersion());
+                    throw new PowerAuthAuthenticationException();
+                }
+
+                // Fetch activation ID and signature type
+                final PowerAuthSignatureTypes signatureFactors = authentication.getSignatureFactors();
+
+                // Fetch data from the request
+                final TokenCreateRequest requestObject = request.getRequestObject();
+                final String ephemeralPublicKey = requestObject.getEphemeralKey();
+                final String encryptedData = requestObject.getEncryptedData();
+                final String mac = requestObject.getMac();
+
+                // Prepare a signature type converter
+                SignatureTypeConverter converter = new SignatureTypeConverter();
+
+                // Get ECIES headers
+                String applicationKey = encryption.getApplicationKey();
+                String activationId = encryption.getActivationId();
+
+                // Create a token
+                final PowerAuthPortV3ServiceStub.CreateTokenResponse token = powerAuthClient.createToken(activationId, applicationKey, ephemeralPublicKey,
+                        encryptedData, mac, converter.convertFrom(signatureFactors));
+
+                // Prepare a response
+                final io.getlime.security.powerauth.rest.api.model.response.v3.TokenCreateResponse responseObject = new io.getlime.security.powerauth.rest.api.model.response.v3.TokenCreateResponse();
+                responseObject.setMac(token.getMac());
+                responseObject.setEncryptedData(token.getEncryptedData());
+                return new ObjectResponse<>(responseObject);
+            } else {
+                throw new PowerAuthAuthenticationException();
+            }
+        }  catch (PowerAuthAuthenticationException ex) {
+            throw ex;
+        } catch (Exception ex) {
+            throw new PowerAuthAuthenticationException(ex.getMessage());
+        }
+
+    }
+
+    @POST
+    @Consumes({MediaType.APPLICATION_JSON})
+    @Produces({MediaType.APPLICATION_JSON})
+    @Path("remove")
+    public ObjectResponse<TokenRemoveResponse> removeToken(ObjectRequest<TokenRemoveRequest> request, @HeaderParam(PowerAuthTokenHttpHeader.HEADER_NAME) String tokenHeader) throws PowerAuthAuthenticationException {
+
+        try {
+            PowerAuthApiAuthentication authentication = authenticationProvider.validateToken(tokenHeader, Arrays.asList(
```

Comment:
Same issue as with create, we need to validate signature...

Reply:
Ok, same fix as for V2.